### PR TITLE
Budget allocation and optimal point estimations

### PR DESCRIPTION
**Pull request by cetagostini**
Hello team!

# Context
As marketers, we're perpetually hunting for ways to amp up the effectiveness of the campaigns we churn out regularly.

Hence, possessing reliable methods to allocate our budget efficiently—wherein we maximize our outcomes—is paramount in ensuring the success of our marketing investments.

## Competitors
Other tools like [Robyn](https://github.com/facebookexperimental/Robyn) and [Lightweight](https://github.com/google/lightweight_mmm) MMM offer features to create budget allocation aimed at maximizing the target. This highlights the importance it holds within the community.

# Goal
Integrating a budget allocator function within PyMC-Marketing that allows users to conveniently select a budget and allocate their resources to those channels that maximize the value of their target variable.

## Secondary goals (Value proposition)
1. Including optional variables that allow the use of business rules within the budget optimizer, so users can tinker more freely and optimize the variable based on their own conceptions (Priors).
2. Incorporating additional information that allows users to better understand where they should not be overspending, and where their spend start to diminish.
3. Add uncertainty around the curves.

# Hypotesis
As we are dealing with diminishing return curves, which have a sigmoidal shape, meaning, as `x` becomes excessively positive (approaching infinity), the sigmoid function approaches its upper limit. **Based on my interpretation of the `logistic_saturation` function on `mmm/transformers.py`.**

We should find different sections on the curve:
1. One where our curve increase and gets more results, even if those results are just a few on absolute numbers.
2. Other where our curve increases very slowly and gets a lot of results on absolute numbers, but at a higher cost.

These two sections of the curve would be divided by the `elbow`, basically where our curve changes direction and starts to be flatted.

image

# Solution
If we know the function that defines the curve of our data, we can project it to discern which channels have the most potential for development according to our model and which do not. This empowers the budget allocator to distribute our resources non-linearly, based on the data.

Example output:
Screenshot 2023-07-25 at 23 15 05

Screenshot 2023-07-25 at 23 09 14

# Work.
To achieve our goal, the first step was to find a function that could interpret the points in a coherent manner. Using the original function was challenging due to its scaled nature, and it operates with values and parameters between 0 and 1, which are difficult to retrieve after the model is trained.

I chose the function of Michael Mentes, which is expressed as follows:
`y = L * x / K + x`
​
where:
- L is the maximum value the curve approaches as x goes to infinity (akin to the saturation level).
- k is the x value when the function reaches half its maximum value. This can be thought of as a measure of the steepness of the curve or the position of the "elbow".

## How look like?
- When `k=0.5`, the curve saturates quickly and then plateaus. This is because the "elbow" is closer to the y-axis.
- As k increases, the curve starts to look more like an linear curve, with the elbow moving closer to the x-axis.
**ps: This given the same value for `L`**

image

## How do these changes affect the current curve fitting?
Since we now have a quadratic function, at times the model tends to believe that after passing the maximum point, any additional input will yield negative returns. This is indeed true when discussing marginal returns, but in this case, our curve is dealing with absolute values.

**Example 1: Current fit**
Screenshot 2023-07-25 at 23 35 19

**Example 2: Modify fit (Michael Mentes)**
Screenshot 2023-07-25 at 23 36 02

Great to hear your opinions here: @ricardoV94 @juanitorduz 

# Questions
-  First of all, I would like to get your feedback on the reasoning and logic behind, it and if makes sense to you to move in this direction with PyMC-Marketing. Based on my experience, makes a lot of sense but I could be biased.
- Even when we use our model to fit the data into a logistic (sigmoid) function, within the response curve plots, we make use of a quadratic (polynomial) fit. Given by default on Seaborn. Why? 

# Help
- **Unit tests,** I already build some but I'm not so familiarized. It would be amazing if you could check the functions built and validate are working as expected.

# Important
This PR is a synthesis of the one previously opened [here](https://github.com/pymc-labs/pymc-marketing/pull/287). The idea is to take all the feedback received earlier to create this cleaner and more succinct draft for new contributors. I hope you find it appropriate. I'm also open to suggestions.

_I still need to run this on some notebooks to ensure that the results are coherent. Although everything has worked as custom functions in my test notebook, I must try installing this version of the PR._

## Code snippet
``` python
#Estimate your curve (Michaelis Menten)
parameters = mmm.compute_channel_estimate_points_original_scale()

#Budget allocation based on the estimations
mmm.budget_allocation(
    total_budget=5,
    parameters = parameters,
    budget_bounds = {'x1':[1,30],
                     'x2':[1,60]
                     }
)
```

``` python
#Check the curve
mmm.plot_direct_contribution_curves(show_estimations=True)
```

## Example notebook
[Google Colab: #329 PR](https://colab.research.google.com/drive/1Q9_5tUczYjH2XQ5OsnesmGssyuu0qS7P?authuser=1#scrollTo=EVdcCGNT7KdB)

cc: @juanitorduz @ricardoV94 @twiecki @cluhmann 

----
:books: Documentation preview :books:: https://pymc-marketing--329.org.readthedocs.build/en/329/

### CODE REVIEW

File: pymc_marketing/mmm/utils.py
Comment (ferrine) [inline pymc_marketing/mmm/utils.py:41]: This is also useful as a dedicated saturation function in mmm.transformers

Comment (cetagostini) [reply pymc_marketing/mmm/utils.py:41]: I like this idea, checking Google MMM lightweight they had included several model configurations (Hill, Carryover, Adstock) which you could choose to define Saturation and Lagging.
**Example:**
Screenshot 2023-08-01 at 14 49 09

They refer to the following wiki where the saturation mentioned it is very similar to the `Michaelis Menten`. I could imagine we can implement something similar here, in order to be more extensive to how different existing data sets can fit better to different saturation and lagging functions depending on their own conditions.

Screenshot 2023-08-01 at 14 46 00

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:486]: Why do you use these 0.99 and 0.01 ?

File: pymc_marketing/mmm/budget_optimizer.py
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py:61]: Here we could also benefit by what you mean by `objective function`

File: pymc_marketing/mmm/budget_optimizer.py
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py:97]: Same comment about `Optional` as above. Also we **should never use mutable structures as defaults**, see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments 😈

**Review by juanitorduz — CHANGES_REQUESTED** [review]
This was very nice and easy o review thanks! I left some small comments about code style. 

I will continue testing the budget allocator in the mmm example notebook :)

File: pymc_marketing/mmm/budget_optimizer.py
Comment (juanitorduz) [reply pymc_marketing/mmm/budget_optimizer.py:97]: Can you please add docstrings for this function? This is very important from a user perspective.

**Comment by juanitorduz** [comment]
@cetagostini Would you mind sharing an example code of how this will be used in our example? You can assume the `mmm` object is already fitted. Thank would help the review. 

I want to run:

```pyhton
mmm.budget_allocation(
    total_budget=1000,
)
```

and I get the (expected error)

```python
TypeError                                 Traceback (most recent call last)
Cell In[43], line 1
----> 1 mmm.budget_allocation(
      2     total_budget=1000,
      3 )

TypeError: BaseMMM.budget_allocation() missing 2 required positional arguments: 'parameters' and 'budget_bounds'
```

So a reproducible code snipped will help the review :)

**Comment by cetagostini** [comment]
@juanitorduz Again, thank you very much for your review!

Before sharing, I did a small verification and found a couple of issues to be corrected. It took me a bit longer, but here is the snippet and a small Colab to check the results, so you can play with it.

## Code snippet
``` python
#Estimate your curve (Michaelis Menten)
parameters = mmm.compute_channel_estimate_points_original_scale()

#Budget allocation based on the estimations
mmm.budget_allocation(
    total_budget=5,
    parameters = parameters,
    budget_bounds = {'x1':[1,30],
                     'x2':[1,60]
                     }
)
```

``` python
#Check the curve
mmm.plot_direct_contribution_curves(show_estimations=True)
```

## Example notebook
[Google Colab: #329 PR](https://colab.research.google.com/drive/1Q9_5tUczYjH2XQ5OsnesmGssyuu0qS7P?authuser=1#scrollTo=EVdcCGNT7KdB)

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:13]: Done, I add a larger docstring trying to describe the process and trying to be specific.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:17]: Copy that!

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:61]: Basically, the negative of the total expected contribution for the given budget allocation.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:101]: I think I fixed it, and removed all possible mutable objects inside my function, but let me know if you see any. The docstring is done as well.

**Comment by cetagostini** [comment]
All changes already done!

**Review by juanitorduz — CHANGES_REQUESTED** [review]

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:550]: Here do you mean optional as in there is a default value (because there is no at the moment)? Recall that the type hint `Optional` refers to the fact that it can take `None` values. Not that is really optional.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py:157]: Again we should remove the `Optional` hint if `budget_ranges` can not be `None`.

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:584]: I would rename this method as `compute_channel_plateat_points_original_scale`

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:603]: can re rename the parameter `show_estimations` to `show_michaelis_menten_fit`

Comment (juanitorduz) [reply pymc_marketing/mmm/base.py:603]: By default we should keep the original range. In the example notebook when I run 

```python
fig = mmm.plot_direct_contribution_curves(show_estimations=True)
[ax.set(xlabel="x") for ax in fig.axes]
```
I get:

image

Comment (juanitorduz) [reply pymc_marketing/mmm/base.py:603]: I would rename `x_stop` to `xlim_max`

Comment (juanitorduz) [reply pymc_marketing/mmm/base.py:603]: Otherwise, using `x_stop = 1.5` I get 
image

File: pymc_marketing/mmm/budget_optimizer.py
Comment (juanitorduz) [inline pymc_marketing/mmm/budget_optimizer.py:169]: From the example notebook I get 
image
Should we add the expected total value instead of `NaN`?

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Looks very nice and works on the example notebook. I left minor comments regarding naming and type hints. After that I think we can merge this an add an expermiental warning to get feedback from users 💪

**Comment by juanitorduz** [comment]
@cetagostini I also suggest opening a followup PR where we load the mmm model from the example (we can use the model builder here) to run an explain this optimization procedure.

**Comment by cetagostini** [comment]
Hey @juanitorduz 

I apply all the changes you mention, and additionally create a new [PR with the documentation here](https://github.com/pymc-labs/pymc-marketing/pull/370).

# **Relevant changes**
## 1 - Adding custom saturation method:
I made some important changes, the first and most important is that I added a new function inside the PR. The idea is that now users can choose the method to fit their curve, not only `michaelis-menten` but also `sigmoid`. For this use the native function of the model with a little adjustment.

$$
\frac{\alpha - \alpha \cdot \exp(-\lambda x)}{1 + \exp(-\lambda x)}
$$

**Where:**
* α (Alpha): Represents the Asymptotic Maximum or Ceiling Value. It is the value the function approaches as the input x becomes large.
* λ (Lambda): Affects the steepness of the curve. A higher value of λ makes the curve steeper, while a lower value makes it more gradual.

**Why modify the function and add the alpha parameter?**
Right now, since most of the process requires understanding the maximum points of the curve and the inflection points, adding it made me more scalable, without requiring so many changes, because now both functions will be estimated with two parameters. Additionally, I think that having a cealing parameter also makes the function somewhat more interpretable, personal opinion.

By the way, and I think we can debate, doing tests I think it would even be good to estimate it directly with the base model. We could have the model with two saturation functions, which could be estimated and chosen by the user. Each one with its pros and cons (But this is the subject of another PR).

**Why not use the internal functions of PyMC?** 
I was playing around with this idea for a while but I didn't get a satisfactory fit. Even passing the base parameters of the model. Also as I mentioned previously, even trying to move forward with this idea, use the function due to the difference in parameters would have required more code lines.

``` python
lam = float(mmm.fit_result['lam'].sel(channel='tiktok_cost_android').mean())
beta = float(mmm.fit_result['beta_channel'].sel(channel='tiktok_cost_android').mean())
alpha = float(mmm.fit_result['alpha'].sel(channel='tiktok_cost_android').mean())

print(beta,lam, alpha)

def max_abs_scaler_factor(x):
    return np.max(np.abs(x))

x = df['tiktok_cost_android'].to_numpy()
y = channel_contributions.sel(channel='tiktok_cost_android').to_numpy()

factor_x = max_abs_scaler_factor(x)
factor_y = max_abs_scaler_factor(y)

x_range = np.linspace(min(x), max(x), 100)

y_saturated = logistic_saturation(
    x=(x_range / factor_x),
    lam=lam
).eval()

plt.scatter(x, y,)
plt.plot(x_range , y_adstock_saturated * factor_y, #re-scaling based on the `y`
linestyle='-')
plt.xlabel('x')
plt.ylabel('y')
plt.title('Fitting Curve')
plt.show()
```
Screenshot 2023-08-29 at 00 04 39

## 2 - Adding plot method to compare budget recommendations:
Basically, based on the recommended budget, create a method to be able to compare the initial scenario with respect to the new ones proposed, allowing you to see the differences.

Here is an small snippet but you can see more on the other PR inside of the [example notebook](https://github.com/pymc-labs/pymc-marketing/pull/370).

``` python
figure_ = mmm.plot_budget_scenearios(base_data=initial_scenario, method='sigmoid', scenarios_data=[result_sigmoid])
```
image

As you can see there are some shaded sections inside the plot, basically I'm looking to add the intervals of the curve using the 0.05 and 0.95 quantiles of our posterior. Meaning, the shaded areas represent the maximum expected contribution, the minimum, and the expected average. 

## 3 - Adding intervals on the method plot_direct_contribution_curves:
Now we can see the intervals, again based on the quantiles and the selected fit (Sigmoid or Menten) in addition to the estimate of the alpha (Ceiling) parameter for both functions.
``` python
sigmoid_response_curve_fig = mmm.plot_direct_contribution_curves(
show_fit = True, method='michaelis-menten'
)
```
image

I would think that with this we are closer to leaving. However, you may be leaving something in the dark regarding hints or unit tests. For which I ask you for help if you consider a function that we should test correctly and you do not see it there in the tests.

I hope for your always appreciated feedback! 🚀

File: pymc_marketing/mmm/base.py
Comment (cetagostini) [reply pymc_marketing/mmm/base.py:682]: Added

File: pymc_marketing/mmm/base.py
Comment (cetagostini) [reply pymc_marketing/mmm/base.py]: Correct, should be optional (None allow). If the user doesn't use it and the value is None then some bounds are used by default.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:180]: Yes, can be optional. If the user doesn't use it and the value is None then some bounds are used by default.

File: pymc_marketing/mmm/base.py
Comment (cetagostini) [reply pymc_marketing/mmm/base.py]: I switch to `compute_channel_curve_parameters_original_scale` given the fact we can use a differents curve fit now. What do you think?

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:200]: Correct, already working as expected.

Screenshot 2023-08-29 at 00 40 31

**Comment by cetagostini** [comment]
# Small changes
1. New docstrings added to secondary support functions. 
2. The error in the unit test was solved (Apparently). Still, I believe the unit test could be basic and need a more sophisticated approach to cover the majority of the cases, a hand here would be nice.
3. Adjust the plot to avoid warning of redundant parameters.

# To be defined
1. Scaled X data: Right now I'm using the parameter `self.X` to perform the fit according to the determined function, the same parameter is used later inside the budget allocation functions, this means that the user would have to pass the scaled budget bounds, and the results of the optimal budget would also be scaled each one to its respective factor according to the channel.
- Easy fix: Replace `self.X` with the unprocessed X data.
- Less easy: Use the internal scaler parameter to unscale the data and return the results to the original scale.

PS: I'm not sure if this was always the case, I think in earlier versions the direct response curves were shown in the original scale. I haven't checked with previous versions, but we should keep an eye on it.

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [reply pymc_marketing/mmm/base.py]: Thanks! Do you want to add a shorter summary to the doc strings? :)

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:520]: Here is some duplicated code. You can use

```python
if method == "michaelis-menten":
    func = estimate_menter_parameters
elif method == "sigmoid":
    func = estimate_sigmoid_parameters

alpha_limit_upper, lam_constant_upper = func(
                channel, self.X, channel_contributions_quantiles.sel(quantile=0.95)
            )
```

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:547]: IMO, this should be in a utils module as it does not depend on the MMM class.

- The type hint should be `keywords: List[str]`

File: pymc_marketing/mmm/base.py
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:909]: I will rename this to `compute_channel_curve_optimization_parameters_original_scale` so that people do not think these are model parameters (adstock or saturation)

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:41]: type hints missing

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:63]: type hints missing

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:51]: what is k? (maybe you add the formula in the doctrings ?)

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:83]: type hints

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:98]: These are not the same as the signature function

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:118]: type hints

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:133]: These names do not correspond to the signature function

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:157]: type hints

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:179]: We can remove this comment

File: pymc_marketing/mmm/utils.py
Comment (juanitorduz) [inline pymc_marketing/mmm/utils.py:189]: type hints

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Looking good! Here are some minor style suggestions

**Comment by cetagostini** [comment]
Unit test from utils working. Still, debugging is needed on the budget optimizer unit test.
Screenshot 2023-09-07 at 01 36 50

**Comment by juanitorduz** [comment]
The lint problem has nothing to do with your code. Is probably and update in mypy, see https://github.com/pymc-labs/pymc-marketing/pull/376/commits/dba627882c70ae5c4c3535f7057073d5ff8734a4

**Comment by cetagostini** [comment]
Budget optimizer unit test:
Screenshot 2023-09-09 at 14 28 14

**Comment by juanitorduz** [comment]
Tests and lint is 🟢 ! Yay! I will take a look in the upcoming days :)

**Comment by twiecki** [comment]
Can we merge?

**Comment by cetagostini** [comment]
> Can we merge?

I think we are set. Not sure if the team has other questions @twiecki 

From my side:
1. We have already all checks except `codecov`. Still, I added a few new unit tests to improve and increase the target.
2. The branch is updated up to September 19th.
3. I ran a test on the current notebook using my branch and everything works properly ([Test here](https://colab.research.google.com/drive/1duJewYhl_lHmv_dO8yavqjYowfqVt0Uj#scrollTo=gvz-1QmP9eXu)).
4. I ran a test on the new example notebook for budget allocation and it works as well ([Test here](https://colab.research.google.com/drive/1r6vryO7h79eM17iTHRsm45ybKybvjWpJ#scrollTo=74q2NZd03SMr))

I added some flexibility to the `curve_fit` functions to give users the possibility to use them outside of the mmm class. This also gives the chance to handle non-fit errors given by the optimizer.

The `pre-commit` looks correct when I test it locally:
Screenshot 2023-09-19 at 20 16 15

The error given now by `docs/readthedocs.org:pymc-marketing` is related to the Numpy new version, but I assume this should be solved from the GitHub config checks. Right?

Wait for your feedback!

cc: @ricardoV94 @juanitorduz

**Comment by juanitorduz** [comment]
Can we add a warning saying that this feature is experimental? What do you think ?

**Comment by cetagostini** [comment]
> Can we add a warning saying that this feature is experimental? What do you think ?

Sure, where do you think can we added? Notebook? @juanitorduz

**Comment by juanitorduz** [comment]
> > Can we add a warning saying that this feature is experimental? What do you think ?
> 
> Sure, where do you think can we added? Notebook? @juanitorduz

Sorry for the late reply 🙈 ! What about the public methods regarding optimization in `pymc_marketing/mmm/base.py`?

**Comment by cetagostini** [comment]
I added an alert to the docstrings located under `base.py`. It is okay? @juanitorduz 

Screenshot 2023-09-30 at 14 23 04

**Comment by juanitorduz** [comment]
> I added an alert to the docstrings located under `base.py`. For each new function, It is okay? @juanitorduz
> 
> Screenshot 2023-09-30 at 14 23 04

Thank you @cetagostini ! In addition, we can add a warning in the code itself as:

```python
import warnings

...

warnings.warn("This budget allocator method is experimental", UserWarning)
```

After that we can merge from my side :)

**Review by juanitorduz — APPROVED** [review]
@cetagostini Thank for for this contribution! IMO let us ship this as experimental and get feedback from users and iterate! 

WDYT @ricardoV94 ? @twiecki ?

**Comment by twiecki** [comment]
Yes, we should merge this.

**Comment by twiecki** [comment]
Incredible contribution @cetagostini!

**Comment by cetagostini** [comment]
Hey guys! Didn't have the time to reply properly a few days ago, just wan to highlight a few things.

1. Really thank you for the support, I'm thrilled to see the merge finally happening 🙌🏻 I'm quite inspired by this momentum, and I'll surely try to help with the other issues that are now active, since this one is closed.
2. A great mention to @juanitorduz without all the time he dedicated and helped me get this PR forward, I would surely have stayed halfway. A great source of inspiration in the community, as well a great person to debate and share ideas reliably in the community.

ps: I'll probably be doing a little self-promotion this week to invite users to test.

Go PyMC-Marketing 🚀

**Pull request merged**
